### DISCUSSION TITLE
March 24 update

### DISCUSSION THREAD

**Discussion by JoernT** [Jinntec/Fore#252] [Show and tell]
# March 24 update
For those interested and for early involvement i plan to give updates here about ongoing development and current plans of what can be expected in the near future.

Would be nice if you give short feedback if this is useful for you. 

## new
* there is a new subproject jinntec/fore-project to kickstart building Fore pages. It comes with a webserver for dev purposes and with End2End testing with Cypress. The repo is a template repository so you can setup your own repo quickly. It's ready to use.

## Currently ongoing developments

* Martin and me are working on DND support in fx-repeat. This has been progressing well so far and we're near merging our efforts into dev branch. Specifically now fx-repeat supports dnd within and between repeats also re-ordering the nodes in the data when you re-arrange the UI. The example which we've made work so far is the demo/kanban.html example. As with existing kanban boards you can re-sort the column entries and move entries between columns. Also the columns itself can be re-arranged which means the dnd works within a nested repeat.
* there's a regression bug with relevance which we are investigating. 
* we started on supporting the `src` attribute on container elements like group, case and repeat which simplifies authoring and gives you more options to assemble your final page from bits and pieces. This involves a bit of refactoring the code to avoid code duplication.
* With our latest changes to shared instances we have made a demo working which separates the model and UI in different files. Again as above this simplifies re-use of page-fragments.

## upcoming

Various tasks and ideas have at least been looked into.

### event hook actions

Always looking for possibilities to reduce markup and simplify certain common patterns and make them easier to understand. Let's face it - plain XForms markup can get convoluted and hard to navigate. By adding logic the markup quickly grows and it gets harder and harder to follow the traces and 'read the page'. 

I came about the idea of 'event hook actions' recently and tried it with one example:
```
<fx-fore>
    <fx-model>
        <fx-action event="model-construct-done">
       .... do some initial stuff ...
        </fx-action>
```

This is a common pattern you need when you need some init for your page. The fx-action allows you to group some actions to do before the UI is up.

My idea was to create alternative elements that make this IMO more speaking by allowing this alternatively:
```
<fx-fore>
    <fx-model>
       <fx-construct-done>
         .... do some initial stuff ...
       </fx-construct-done>
```

I implemented that and it was more than trivial.

Idea would be to offer those elements for the most common situations like:
* <fx-submit> as alternative to <fx-action event="submit">
* <fx-submit-done> as alternative to <fx-action event="submit-done">
* <fx-submit-error> as alternative to <fx-action event="submit-errror">
* <fx-value-changed> as alternative to <fx-action event="value-changed">

Of course all these further support the 'if', 'delay' and 'while' stuff as fx-action does cause they simply inherit from it.

IMO these also improve the readability as you also see where the action ends - so e.g. </fx-construct-done> at the end of the block instead of </fx-action> which doesn't tell you which event handler ends here.

I appreciate any feedback on this topic but cost of implementation is rather low.

### iterate

Implementation already started but stalled due to time constraints. Definitely high on list.

### upload

upload can be done with usual HTML within Fore but this does not allow to associate the meta-data you probably like to set the uploaded content into context. This still need integration with submisssion directly. 

### fx-connection

Again - already started the implementation but lacking capacities currently to get going but i'm optimistic we'll that later this year. This will be a bit more effort to provide a stable and well-tested implementation.

This is it for today. Try to give more regular updates at least once a month and we appreciate any input, request or suggestion. Even you feel you couldn't contribute you may still contribute - by dropping feedback, providing or requesting examples. We learn from practice and improve the product on the go.

**Comment by timathom** [comment]
Thanks, @JoernT, this is very useful! It's exciting to see that DnD functionality is coming. Another thing you might add to the list is the support for [E2E testing](https://github.com/Jinntec/Fore/discussions/248) for form developers via Cypress.

**Reply by JoernT** [reply]
@timathom what do you have in mind with cypress testing? Actually you can already do that. Either when just cloning Fore and adding files to cypress/e2e folder or by using fore-project and essentially set up a project that is separate from Fore (recommended) and only contains what you're intereste in. 

With fore-project you have Cypress been setup and you can start right out of the box. Maybe some samples from the Fore repo would be helpful as starters. Maybe it needs some more recipes for certain things?

If you got a different setup somehow please let me know.

**Reply by JoernT** [reply]
On another track - it would be good to learn about your concrete use case with DnD. Can you give a short description of what you'd like to achieve or even can post a screenshot? Just to make sure we cover expectations in the end.

**Reply by timathom** [reply]
@JoernT Re: Cypress, I only meant that you might want to highlight that functionality (i.e., fore-project) in your update as something that Fore developers could take advantage of :)

I'll think more about the DnD use case.

**Reply by JoernT** [reply]
@timathom good point - should stand out somewhere a bit more
